server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/AllocationActionMultiListener.java

```diff
@@ -30,7 +29,7 @@ public AllocationActionMultiListener(ThreadContext context) {
     }
 
     public ActionListener<T> delay(ActionListener<T> delegate) {
-        final var wrappedDelegate = wrapPreservingContext(delegate, context);
+        final var wrappedDelegate = new ContextPreservingActionListener<>(context.newRestorableContext(false), delegate);
```

Comment:
I believe this was changed because `wrapPreservingContext` sets `preserveResponseHeaders=true`.
Should we have overloaded `wrapPreservingContext` that allows to change this flag instead?

Reply:
We probably should, this is a little trappy. I'll not do it here tho, it's quite a big change, but feel free to follow up.